core/TypeErrorDiagnostics.cc

```diff
@@ -134,4 +135,99 @@ void TypeErrorDiagnostics::explainUntyped(const GlobalState &gs, ErrorBuilder &e
     }
 }
 
+namespace {
+optional<core::AutocorrectSuggestion::Edit> autocorrectEditForDSLMethod(const GlobalState &gs, Loc insertLoc,
+                                                                        string_view prefix, ClassOrModuleRef dslOwner,
+                                                                        string_view dsl, bool needsDslOwner) {
+    if (needsDslOwner) {
+        if (dsl == "") {
+            return core::AutocorrectSuggestion::Edit{
+                insertLoc,
+                fmt::format("{}extend {}\n", prefix, dslOwner.show(gs)),
+            };
+        } else {
+            return core::AutocorrectSuggestion::Edit{
+                insertLoc,
+                fmt::format("{}extend {}\n{}{}\n", prefix, dslOwner.show(gs), prefix, dsl),
+            };
+        }
+    } else if (dsl != "") {
+        return core::AutocorrectSuggestion::Edit{
+            insertLoc,
+            fmt::format("{}{}\n", prefix, dsl),
+        };
+    } else {
+        return nullopt;
+    }
+}
+} // namespace
+
+// dslOwner can be noClassOrModule() to simply unconditionally insert the `dsl` string
+// dsl can be `""` to simply insert `extend {dslOwner}` if dslOwner is not already an ancestor
+optional<core::AutocorrectSuggestion::Edit>
+TypeErrorDiagnostics::editForDSLMethod(const GlobalState &gs, FileRef fileToEdit, Loc defaultInsertLoc,
+                                       ClassOrModuleRef inWhatRef, ClassOrModuleRef dslOwner, string_view dsl) {
+    auto inWhat = inWhatRef.data(gs);
+    auto inWhatSingleton = inWhat->lookupSingletonClass(gs);
+
+    auto needsDslOwner = false;
+    if (dslOwner.exists()) {
+        needsDslOwner = !inWhatSingleton.data(gs)->derivesFrom(gs, dslOwner);
+    }
+
+    auto inCurrentFile = [&](const auto &loc) { return loc.file() == fileToEdit; };
```

Comment:
[nit, feel free to ignore]
Would it be possible to capture only `fileToEdit` in  `inCurrentFile`?
```suggestion
    auto inCurrentFile = [&fileToEdit](const auto &loc) { return loc.file() == fileToEdit; };
```

Reply:
If the lambda were longer, I would definitely agree with you. As is, I personally feel like the lambda is short enough to warrant the `[&]` capture. In some other places we use explicit capturing where the lambda is more complicated, we prefer explicit captures.

(Given how short it is, I feel like it would just be one more thing to forget to change if this code were to be edited. As in, you'd rename the variable, or change the `==` condition, but forget to change the list of captures and get a toilsome compiler error.)